File: src/contests/cg/puzzle/scala/telephone-numbers.scala
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

import scala.collection.mutable

object Solution {
  def main(args: Array[String]): Unit = {
    val n = scala.io.StdIn.readLine().trim.toInt
    val root = mutable.Map[Char, Any]()
    var cable = 0
    for (_ <- 0 until n) {
      val number = scala.io.StdIn.readLine().trim
      var node = root
      for (d <- number) {
        if (!node.contains(d)) {
          node(d) = mutable.Map[Char, Any]()
          cable += 1
        }
        node = node(d).asInstanceOf[mutable.Map[Char, Any]]
      }
    }
    println(cable)
  }
}

File: src/contests/cg/puzzle/perl/telephone-numbers.pl
# 🎮 CodinGame Puzzle - telephone-numbers
# https://www.codingame.com/training/medium/telephone-numbers

my $n = <STDIN>;
chomp $n;
my %root;
my $cable = 0;
for (1 .. $n) {
    my $number = <STDIN>;
    $number =~ s/\s+$//;
    my $node = \%root;
    for my $d (split //, $number) {
        if (!exists $node->{$d}) {
            $node->{$d} = {};
            $cable++;
        }
        $node = $node->{$d};
    }
}
print "$cable\n";
